chanotify/chanotify.go

```diff
@@ -17,85 +11,68 @@ type dataPair struct {
 // notification from Notifier, close doesn't spawn anything and removes channel
 // from Notifier.
 type Notifier struct {
-	c     chan string
-	chMap map[<-chan struct{}]*dataPair
-	ctrl  chan struct{}
-	m     sync.Mutex
+	c chan string
+
+	m      sync.Mutex // guards doneCh
+	doneCh map[string]chan struct{}
 }
 
-// New returns already running *Notifier.
+// New returns a new *Notifier.
 func New() *Notifier {
 	s := &Notifier{
-		c:     make(chan string),
-		chMap: make(map[<-chan struct{}]*dataPair),
-		ctrl:  make(chan struct{}),
+		c:      make(chan string),
+		doneCh: make(map[string]chan struct{}),
 	}
-	go s.start()
 	return s
 }
 
 // Chan returns channel on which client listen for notifications.
-// Ids of notifications is sent to that channel.
+// IDs of notifications is sent to the returned channel.
 func (s *Notifier) Chan() <-chan string {
 	return s.c
 }
 
-// Add adds new notification channel to Notifier.
-func (s *Notifier) Add(ch <-chan struct{}, id string) {
+func (s *Notifier) killWorker(id string, done chan struct{}) {
 	s.m.Lock()
-	s.chMap[ch] = &dataPair{
-		id: id,
-		selectCase: reflect.SelectCase{
-			Dir:  reflect.SelectRecv,
-			Chan: reflect.ValueOf(ch),
-		},
-	}
+	delete(s.doneCh, id)
 	s.m.Unlock()
-	s.ctrl <- struct{}{}
+	close(done)
```

Comment:
Why do you need to close done here? As I see there won't be any reads from it at this point.